File: LeetCode/0033/golang/lc0033.go
package lc0033

// Runtime complexity: O(log(L))
// Auxiliary space: O(1)
// Subjective level: medium-
// Solved on: 2026-03-26
func search(nums []int, target int) int {
	// 1. Start the search with full range in consideration:
	L := len(nums)
	left := 0
	right := L - 1

	// 2. Divide the range in half. Determine whether each partition
	// contains a pivot (it does when subrange_left > subrange_right).
	for left < right {
		mid := left + (right-left)/2
		if nums[left] <= nums[mid] {
			// handle the case of left partition not containing a pivot point.
			// does it contain the target?
			if nums[left] <= target && target <= nums[mid] {
				// the target is contained within the left partition exclusively;
				// continue searching there.
				right = mid
			} else {
				// the target is contained within the right partition exclusively
				// (the one with a pivot point). Continue searching there.
				left = mid + 1
			}
		} else {
			// handle the case of left partition containing a pivot point;
			// the code is similar to the one handling pivot point in the right partition.
			// Does the pivot-free right partition contain the target?
			if nums[mid+1] <= target && target <= nums[right] {
				// the target is contained within the right partition exclusively;
				// continue searching there.
				left = mid + 1
			} else {
				// the target is contained within the left partition (the one with
				// a pivot point). Continue searching there.
				right = mid
			}
		}
	}

	// left now either points to the searched for target;
	// or target is entirely not present in nums.
	if nums[left] == target {
		return left
	}
	return -1
}

File: LeetCode/0033/python/lc0033.py
from typing import List

# Re-solved in Python, based on my Golang solution from 2026-03-26.

# Runtime complexity: O(log(n))
# Auxiliary space complexity: O(1)
# Subjective level: easy+
# Solved on: 2026-05-22
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        L = len(nums)
        left, right = 0, L-1

        while left < right:
            # Divide the range in half; find the part that contains a pivot
            # (pivot present when subrange_left > subrange_right).
            mid = (left+right) // 2
            if nums[left] <= nums[mid]:
                # left part does not contain a pivot point
                if nums[left] <= target and target <= nums[mid]:
                    # the target is contained somewhere within the left part. Continue searching there.
                    right = mid
                else:
                    # the target is contained somewhere within the right partition.
                    # Continue search there.
                    left = mid+1
            else:
                # left part contains a pivot point.
                if nums[mid+1] <= target and target <= nums[right]:
                    left = mid+1
                else:
                    right = mid
        return left if nums[left] == target else -1
